File: plaoc/src/emulator/helper/StateObservable.ts
import { Ipc, createSignal, mapHelper, simpleEncoder } from "../../../deps.ts";
import { $Callback, $OffListener } from "../../client/helper/createSignal.ts";

export class StateObservable {
  constructor(readonly getStateJson: () => string) {}

  private _observerIpcMap = new Map<Ipc, $OffListener>();

  private _changeSignal = createSignal();

  private _observe = (cb: $Callback) => this._changeSignal.listen(cb);

  private _controllersMap = new Map<Ipc, ReadableStreamDefaultController>();

  startObserve(ipc: Ipc, controller: ReadableStreamDefaultController) {
    this._controllersMap.set(ipc, controller);
    mapHelper.getOrPut(this._observerIpcMap, ipc, () => {
      return this._observe(() => {
        controller?.enqueue(simpleEncoder(this.getStateJson() + "\n", "utf8"));
      });
    });
  }
  notifyObserver() {
    this._changeSignal.emit();
  }
  stopObserve(ipc: Ipc) {
    const controller = this._controllersMap.get(ipc);
    if (controller === undefined) throw new Error(`controller === undefined`);
    controller.close();
    this._controllersMap.delete(ipc);
    return mapHelper.getAndRemove(this._observerIpcMap, ipc)?.apply(undefined);
  }
}

File: android/app/src/main/java/info/bagen/dwebbrowser/microService/sys/helper/StateObservable.kt
package info.bagen.dwebbrowser.microService.sys.helper

import org.dweb_browser.microservice.ipc.Ipc
import org.dweb_browser.microservice.ipc.helper.IpcEvent
import info.bagen.dwebbrowser.util.IsChange
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.dweb_browser.helper.*
import org.dweb_browser.microservice.ipc.helper.ReadableStream
import java.io.InputStream

open class StateObservable(
  needFirstCall: Boolean = false,
  private val getStateJson: () -> String,
) {
  val stateChanges = IsChange(needFirstCall)
  private val changeSignal = Signal<String>()
  fun observe(cb: Callback<String>) = changeSignal.listen(cb)

  private val observerReadableSteam = mutableMapOf<Ipc, ReadableStream.ReadableStreamController>()
  suspend fun startObserve(ipc: Ipc): InputStream {
    return ReadableStream(onStart = { controller ->
      observerReadableSteam.getOrPut(ipc) {
        controller
      }
      observe { state ->
        try {
          withContext(Dispatchers.IO) {
            controller.enqueue((state + "\n").toByteArray())
          }
        } catch (e: Exception) {
          controller.close()
          e.printStackTrace()
        }
      }
    })
  }

  fun notifyObserver() {
    runBlockingCatching {
      changeSignal.emit(getStateJson())
    }.getOrNull()
  }

  fun stopObserve(ipc: Ipc) = observerReadableSteam.remove(ipc)?.let { off ->
    off.close()
    true
  } ?: false
}